shell/client/grain-client.js

```diff
@@ -2037,3 +2038,117 @@ Router.map(function () {
     },
   });
 });
+
+Meteor.startup(function () {
+  if (isStandalone()) {
+    // TODO(soon): delete all other routes (maybe change how they're created?)
+    const route = Router.routes.root;
+    _.extend(route.options, {
+      template: "grain",
+      loadingTemplate: "loadingNoMessage",
+
+      waitOn: function () {
+        const standalone = globalDb.collections.standaloneDomains.findOne({
+          _id: window.location.hostname, });
+
+        const subs = [
+          Meteor.subscribe("devPackages"),
+          Meteor.subscribe("standaloneDomain", window.location.hostname),
+          Meteor.subscribe("grainsMenu"),
+          // This subscription gives us the data we need for deciding whether to automatically reveal
+          // our identity.
+          // TODO(soon): Subscribe to contacts instead.
+        ];
+        if (standalone) {
+          subs.push(Meteor.subscribe("tokenInfo", standalone.token));
+        }
+
+        return subs;
+      },
+
+      onBeforeAction: function () {
+        // Don't do anything for non-account users.
+        if (Meteor.userId() && !Meteor.user().loginIdentities) return;
+
+        // Only run the hook once.
+        if (this.state.get("beforeActionHookRan")) return this.next();
+        this.state.set("beforeActionHookRan", true);
+
+        Tracker.nonreactive(() => {
+          const token = globalDb.collections.standaloneDomains.findOne({
+            _id: window.location.hostname, }).token;
+          const path = "/" + (this.params.path || "") + (this.originalUrl.match(/[#?].*$/) || "");
+          const hash = this.params.hash;
+
+          const tokenInfo = TokenInfo.findOne({ _id: token });
+          if (!tokenInfo) {
+            return this.next();
+          } else if (tokenInfo.invalidToken) {
+            this.state.set("invalidToken", true);
+          } else if (tokenInfo.revoked) {
+            this.state.set("revoked", true);
+          } else if (tokenInfo.identityOwner) {
+            this.state.set("invalidToken", true);
+          } else if (tokenInfo.alreadyRedeemed) {
```

Comment:
This is an incredibly weird code path to hit, since you would have to have visited the grain using exactly the same webkey as the standalone domain. Having been shared the grain in any other manner wouldn't trigger this.

That being said, it's obviously a bug. I'll go ahead and add a parameter to the tokenInfo sub.